kiwipy/communications.py

```diff
@@ -129,26 +137,31 @@ def remove_rpc_subscriber(self, identifier):
         except KeyError:
             raise ValueError("Unknown subscriber '{}'".format(identifier))
 
-    def add_task_subscriber(self, subscriber):
+    def add_task_subscriber(self, subscriber, identifier=None):
         """
         Register a task subscriber
 
         :param subscriber: The task callback function
+        :param identifier: the subscriber identifier
         """
         self._ensure_open()
-        self._task_subscribers.append(subscriber)
+        identifier = identifier or shortuuid.uuid()
+        if identifier in self._rpc_subscribers:
+            raise exceptions.DuplicateSubscriberIdentifier("RPC identifier '{}'".format(identifier))
+        self._task_subscribers[identifier] = subscriber
+        return identifier
 
-    def remove_task_subscriber(self, subscriber):
+    def remove_task_subscriber(self, identifier):
         """
         Remove a task subscriber
 
-        :param subscriber: The task callback function
+        :param identifier: The task subscriber identifier
         """
         self._ensure_open()
         try:
-            self._task_subscribers.remove(subscriber)
+            self._task_subscribers.pop(identifier)
```

Comment:
this should raise a `KeyError` not a `ValueError` because `_task_subscribers` is now a dictionary. Should probably add a test for `remote_task_subscriber` to test this behavior

Reply:
This was a inconsistent but I've added a test to check both communicators.  IMHO this shouldn't be a `KeyError` because the fact that it's stored in a `dict` is an implementation detail.  I've consistently made it a `ValueError`.

Reply:
My comment was not fully clear, my point is not that your method should raise a `KeyError`, I agree a `ValueError` is better. However, I was pointing out that the line `self._task_subscribers.pop(identifier)` will raise a `KeyError` so you should catch _that_ and not a `ValueError`

Reply:
Ah yes, good catch.  That last commit changed it and the test now checks for `ValueError` being raised at the API level.